evalml/tests/pipeline_tests/test_component_graph.py

```diff
@@ -0,0 +1,332 @@
+    assert component_graph.get_last_component() is None
+
+
+@patch('evalml.pipelines.components.Transformer.fit_transform')
+@patch('evalml.pipelines.components.Estimator.fit')
+@patch('evalml.pipelines.components.Estimator.predict')
+def test_fit(mock_fit_transform, mock_fit, mock_predict, example_graph, X_y_binary):
+    X, y = X_y_binary
+    mock_fit_transform.return_value = pd.DataFrame(X)
+    mock_fit.return_value = Estimator
+    mock_predict.return_value = pd.Series(y)
+    component_graph = ComponentGraph(example_graph).instantiate({})
+    component_graph.fit(X, y)
+
+    assert mock_fit_transform.call_count == 3
+    assert mock_fit.call_count == 3
+    assert mock_predict.call_count == 3
+
+
+@patch('evalml.pipelines.components.Transformer.fit_transform')
+@patch('evalml.pipelines.components.Estimator.fit')
+@patch('evalml.pipelines.components.Estimator.predict')
+def test_fit_features(mock_predict, mock_fit, mock_fit_transform, X_y_binary):
+    X, y = X_y_binary
+    component_list = ['Imputer', 'One Hot Encoder', 'Random Forest Classifier']
+    component_graph = ComponentGraph.from_list(component_list)
+    component_graph.instantiate({})
+
+    mock_X_t = pd.DataFrame(np.ones(pd.DataFrame(X).shape))
+    mock_fit_transform.return_value = mock_X_t
+    mock_fit.return_value = Estimator
+    mock_predict.return_value = pd.Series(y)
+
+    X_t = component_graph.fit_features(X, y)
+
+    assert mock_fit_transform.call_count == 2
+    assert mock_fit.call_count == 0
+    assert mock_predict.call_count == 0
+
+    pd.testing.assert_frame_equal(X_t, mock_X_t)
+
+
+@patch('evalml.pipelines.components.Transformer.transform')
+@patch('evalml.pipelines.components.Estimator.fit')
+@patch('evalml.pipelines.components.Estimator.predict')
+def test_predict(mock_transform, mock_fit, mock_predict, example_graph, X_y_binary):
+    X, y = X_y_binary
+    mock_transform.return_value = pd.DataFrame(X)
+    mock_fit.return_value = Estimator
+    mock_predict.return_value = pd.Series(y)
+    component_graph = ComponentGraph(example_graph).instantiate({})
+    component_graph.fit(X, y)
+
+    component_graph.predict(X)
+    assert mock_transform.call_count == 6  # Called thrice when fitting pipeline, thrice when predicting
+    assert mock_fit.call_count == 3  # Only called during fit, not predict
+
+
+@patch('evalml.pipelines.components.Transformer.transform')
+@patch('evalml.pipelines.components.Estimator.fit')
+@patch('evalml.pipelines.components.Estimator.predict')
+def test_transform_features(mock_transform, mock_fit, mock_predict, example_graph, X_y_binary):
+    X, y = X_y_binary
+    mock_transform.return_value = pd.DataFrame(X)
+    mock_fit.return_value = Estimator
+    mock_predict.return_value = pd.Series(y)
+    component_graph = ComponentGraph(example_graph).instantiate({})
+    component_graph.fit(X, y)
+
+    component_graph.transform_features(X)
+    assert mock_transform.call_count == 5
+    assert mock_fit.call_count == 3
```

Comment:
@eccabay these mock tests are pretty cool!

These mock tests ensure that the methods were called the right number of times and that the final component's predict/transform was called. But there's still a need for a test which checks that the correct data is passed from component to component. 

I wrote up a test which would cover this. My syntax is probably not perfect, and I may not be 100% about the expected values at the end, but I can't be far off! 😅  Lmk if you wanna chat about it.

```python
@pytest.fixture
def dummy_components():
    class DummyTransformer(Transformer):
        def __init__(self, parameters={}, random_state=0):
            super().__init__(parameters=parameters, component_obj=None, random_state=random_state)
        def fit(self, X, y):
            return self
        def transform(self, X, y=None):
            return pd.DataFrame(np.zeros(X.shape))

    class TransformerA(DummyTransformer):
        """copy class"""
    class TransformerB(DummyTransformer):
        """copy class"""
    class TransformerC(DummyTransformer):
        """copy class"""

    class DummyEstimator(Estimator):
        def __init__(self, parameters={}, random_state=0):
            super().__init__(parameters=parameters, component_obj=None, random_state=random_state)
        def fit(self, X, y):
            return self
        def predict(self, X):
            return pd.Series(np.zeros(len(X)))

    class EstimatorA(DummyEstimator):
        """copy class"""
    class EstimatorB(DummyEstimator):
        """copy class"""
    class EstimatorC(DummyEstimator):
        """copy class"""

    return TransformerA, TransformerB, TransformerC, EstimatorA, EstimatorB, EstimatorC

@patch('TransformerC.transform')
@patch('TransformerB.transform')
@patch('TransformerA.transform')
@patch('EstimatorC.predict')
@patch('EstimatorB.predict')
@patch('EstimatorA.predict')
def test_component_graph_evaluation_plumbing(mock_transa, mock_transb, mock_transc, mock_preda, mock_predb, mock_predc, dummy_components):
    mock_transa.return_value = pd.DataFrame({'feature trans': [1, 0, 0, 0, 0, 0], 'feature a': np.ones(6))
    mock_transb.return_value = pd.DataFrame({'feature trans': [0, 1, 0, 0, 0, 0], 'feature b': np.ones(6) * 2)
    mock_transc.return_value = pd.DataFrame({'feature trans': [0, 0, 1, 0, 0, 0], 'feature c': np.ones(6) * 3)
    mock_preda.return_value = pd.Series([0, 0, 0, 1, 0, 0])
    mock_predb.return_value = pd.Series([0, 0, 0, 0, 1, 0])
    mock_predc.return_value = pd.Series([0, 0, 0, 0, 0, 1])
    TransformerA, TransformerB, TransformerC, EstimatorA, EstimatorB, EstimatorC = dummy_components
    graph = {
        'transformer a': [TransformerA],
        'transformer b': [TransformerB, 'transformer a'],
        'transformer c': [TransformerC, 'transformer a', 'transformer b']
        'estimator a': [EstimatorA],
        'estimator b': [EstimatorB, 'transformer a'],
        'estimator c': [EstimatorC, 'transformer a', 'estimator a', 'transformer b', 'estimator b', 'transformer c']
    }
    component_graph = ComponentGraph(graph)
    component_graph.instantiate({})
    X = pd.DataFrame({'feature1': np.zeros(6), 'feature2': np.zeros(6))
    y = pd.Series(np.zeros(6))
    component_graph.fit(X, y)
    predict_out = component_graph.predict(X)

    pandas.test.assert_frame_equal(mock_transa.call_args[0], X)
    pandas.test.assert_frame_equal(mock_transb.call_args[0], pd.DataFrame({'feature trans': [1, 0, 0, 0, 0, 0], 'feature a': np.ones(6)}, columns=['feature trans', 'feature a'])) # added columns to these to pin the column order
    pandas.test.assert_frame_equal(mock_transc.call_args[0], pd.DataFrame({'feature trans': [1, 0, 0, 0, 0, 0], 'feature a': np.ones(6), 'feature b': np.ones(6) * 2}, columns=['feature trans', 'feature a', 'feature b']))
    pandas.test.assert_frame_equal(mock_preda.call_args[0], X)
    pandas.test.assert_frame_equal(mock_predb.call_args[0], pd.DataFrame({'feature trans': [1, 0, 0, 0, 0, 0], 'feature a': np.ones(6)}, columns=['feature trans', 'feature a']))
    pandas.test.assert_frame_equal(mock_predc.call_args[0], pd.DataFrame({'feature trans': [1, 0, 0, 0, 0, 0], 'feature a': np.ones(6), 'estimator a.y': [0, 0, 0, 1, 0, 0], 'feature b': np.ones(6) * 2, 'estimator b.y': [0, 0, 0, 0, 1, 0], 'transformer c': np.ones(6) * 3}, columns=['feature trans', 'feature a', 'estimator a.y', 'feature b', 'estimator b.y', 'feature c']))
    pandas.test_assert_series_equal(predict_out, pd.Series([0, 0, 0, 0, 0, 1]))
```